### PR TITLE
Fix circular recursion issue in hipification 

### PR DESCRIPTION
**Pull request by lcskrishna**
This PR fixes the circular issue during the hipification process by using Iterative DFS instead of DFS with the introduction of current_state. Also, a new class HipifyResult has been introduced which holds hipified_path, status and current_state for each individual source file instead of a dict.

### CODE REVIEW

File: torch/utils/hipify/hipify_python.py [hunk 1]
```diff
@@ -895,14 +936,22 @@ def repl(m):
         try:
             with clean_ctx.open(fout_path, 'w', encoding='utf-8') as fout:
                 fout.write(output_source)
-            return {"hipified_path": fout_path, "status": "[ok]"}
+            hipify_result.hipified_path = fout_path
+            hipify_result.status = "[ok]"
+            hipify_result.current_state = "done"
```
Comment (jithunnair-amd) [inline torch/utils/hipify/hipify_python.py:941]: Let's replace these strings with an enum or equivalent.

File: torch/utils/hipify/hipify_python.py [hunk 2]
```diff
@@ -844,6 +866,20 @@ def repl(m):
...
+                        header_fout_path = os.path.abspath(os.path.join(output_directory, get_hip_file_path(header_rel_path, is_pytorch_extension)))
+                        header_result.hipified_path = header_fout_path
+                        HIPIFY_FINAL_RESULT[header_filepath] = header_result
+                        return templ.format(os.path.relpath(header_fout_path if header_fout_path is not None
+                                                    else header_filepath, header_dir))
+                    elif header_result.current_state == "done":
+                        hipified_header_filepath = HIPIFY_FINAL_RESULT[header_filepath].hipified_path
+                        return templ.format(os.path.relpath(hipified_header_filepath if hipified_header_filepath is not None
+                                                    else header_filepath, header_dir))
                 hipified_header_filepath = HIPIFY_FINAL_RESULT[header_filepath]["hipified_path"]
```
Comment (jithunnair-amd) [inline torch/utils/hipify/hipify_python.py:883]: We can probably remove these

File: torch/utils/hipify/hipify_python.py [hunk 3]
```diff
@@ -854,6 +890,8 @@ def repl(m):
     output_source = RE_ANGLE_HEADER.sub(mk_repl('#include <{0}>', False), output_source)
     output_source = RE_THC_GENERIC_FILE.sub(mk_repl('#define THC_GENERIC_FILE "{0}"'), output_source)
 
+    hipify_result = HIPIFY_FINAL_RESULT[fin_path]
```
Comment (jithunnair-amd) [inline torch/utils/hipify/hipify_python.py:893]: Please add a comment that this is to get the updated value of `HIPIFY_FINAL_RESULT[fin_path]`

**Review by jithunnair-amd — CHANGES_REQUESTED** [review]
Looks good, needs minor changes

**Comment by jithunnair-amd** [comment]
Let's mention some details of the flash_attention hipification issue that you resolved using this PR

File: torch/utils/hipify/hipify_python.py (hunk 1 above)
Comment (lcskrishna) [reply torch/utils/hipify/hipify_python.py]: Done.

**Comment by lcskrishna** [comment]
> Let's mention some details of the flash_attention hipification issue that you resolved using this PR

The circular issue arises when we try to hipify flash-attention_private include dirs from these two files which import each other.
```
https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/710e20819c9065c58610d4ad11a0eea2c19db91e/include/ck/utility/multi_index.hpp
https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/710e20819c9065c58610d4ad11a0eea2c19db91e/include/ck/utility/common_header.hpp
```

The following error is seen:
```
File "/opt/conda/lib/python3.8/site-packages/hipify_torch/hipify_python.py", line 822, in preprocessor
    output_source = RE_QUOTE_HEADER.sub(mk_repl('#include "{0}"', True), output_source)
  File "/opt/conda/lib/python3.8/site-packages/hipify_torch/hipify_python.py", line 812, in repl
    preprocess_file_and_save_result(output_directory,
  File "/opt/conda/lib/python3.8/site-packages/hipify_torch/hipify_python.py", line 177, in preprocess_file_and_save_result
    result = preprocessor(output_directory, filepath, all_files, header_include_dirs, stats,
  File "/opt/conda/lib/python3.8/site-packages/hipify_torch/hipify_python.py", line 822, in preprocessor
    output_source = RE_QUOTE_HEADER.sub(mk_repl('#include "{0}"', True), output_source)
  File "/opt/conda/lib/python3.8/site-packages/hipify_torch/hipify_python.py", line 812, in repl
    preprocess_file_and_save_result(output_directory,
  File "/opt/conda/lib/python3.8/site-packages/hipify_torch/hipify_python.py", line 177, in preprocess_file_and_save_result
    result = preprocessor(output_directory, filepath, all_files, header_include_dirs, stats,
  File "/opt/conda/lib/python3.8/site-packages/hipify_torch/hipify_python.py", line 733, in preprocessor
    rel_filepath = os.path.relpath(filepath, output_directory)
  File "/opt/conda/lib/python3.8/posixpath.py", line 471, in relpath
    start_list = [x for x in abspath(start).split(sep) if x]
  File "/opt/conda/lib/python3.8/posixpath.py", line 375, in abspath
    if not isabs(path):
  File "/opt/conda/lib/python3.8/posixpath.py", line 63, in isabs
    sep = _get_sep(s)
  File "/opt/conda/lib/python3.8/posixpath.py", line 42, in _get_sep
    if isinstance(path, bytes):
RecursionError: maximum recursion depth exceeded while calling a Python object
```

This PR fixes the recursion issue and hipifies successfully.

**Comment by jithunnair-amd** [comment]
@lcskrishna Let's file an upstream PR for this

**Comment by lcskrishna** [comment]
Created an upstream PR: https://github.com/pytorch/pytorch/pull/104085

**Pull request merged**